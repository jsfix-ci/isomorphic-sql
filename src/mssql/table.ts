import { toForeignKeyConstraint, toPrimaryKeyConstraint, toUniqueConstraint } from ".";

export interface IColumn {
    name: string;
    dataType: string;
    primaryKey?: boolean;
    identity?: number[];
    nullable: boolean;
}

export interface IColumnOptions {
    primaryKey?: IColumn['primaryKey'];
    identity?: IColumn['identity'];
    nullable?: IColumn['nullable'];
}

const tab = '  ';

export class Table {
    private columns: IColumn[] = [];
    private constraints: string[] = [];

    private constructor(public name: string) { }

    static named(name: string) {
        return new Table(name);
    }

    column(name: string, dataType: string, options?: IColumnOptions) {
        const { primaryKey, identity, nullable } = options || {};
        this.columns.push({ name, dataType, primaryKey, identity, nullable });
        return this;
    }

    pk(...columns: string[]) {
        this.constraints.push(toPrimaryKeyConstraint(this.name, columns));
        return this;
    }

    fk(column: string, targetTable: string, targetColumn: string) {
        this.constraints.push(toForeignKeyConstraint(this.name, column, targetTable, targetColumn));
        return this;
    }

    unique(...columns: string[]) {
        this.constraints.push(toUniqueConstraint(this.name, columns));
        return this;
    }

    create() {
        return `CREATE TABLE ${this.name} (
${this.columns.map(c =>
            `${tab}${c.name} ${c.dataType} ${c.primaryKey ? 'PRIMARY KEY' : ''} ${c.identity ? `IDENTITY(${c.identity.join(', ')})` : ''} ${c.nullable ? 'NULL' : 'NOT NULL'}`
        ).join(',\n')}
);
        
${this.constraints.join(';\n')}
`;
    }
}