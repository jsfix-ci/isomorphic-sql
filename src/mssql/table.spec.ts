import { Table } from './table';

describe(Table.name, () => {
    it('scripts a table definition with no constraints', () => {
        const table = Table.named('food')
            .column('id', 'int', {identity: [1,1], primaryKey: true})
            .column('name', 'nvarchar(255)')
            .column('inventory_id', 'int', { nullable: true });            

        expect(table.create()).toMatchSnapshot();
    });

    it('scripts a table definition with constraints', () => {
        const table = Table.named('food')
            .column('id', 'int')
            .column('name', 'nvarchar(255)')
            .column('plu', 'nvarchar(255)')
            .column('inventory_id', 'int')
            .pk('id')
            .fk('inventory_id', 'inventory', 'id')
            .unique('name', 'plu');

        expect(table.create()).toMatchSnapshot();
    });
});