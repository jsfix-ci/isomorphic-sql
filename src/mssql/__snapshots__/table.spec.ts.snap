// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table scripts a table definition with constraints 1`] = `
"CREATE TABLE food (
  id int   NOT NULL,
  name nvarchar(255)   NOT NULL,
  plu nvarchar(255)   NOT NULL,
  inventory_id int   NOT NULL
);
        
ALTER TABLE food ADD CONSTRAINT PK_food_id 
PRIMARY KEY (id);
ALTER TABLE food ADD CONSTRAINT FK_food_inventory_id_inventory_id 
FOREIGN KEY inventory_id REFERENCES inventory(id);
ALTER TABLE food ADD CONSTRAINT IX_food_name_plu_unique UNIQUE(name, plu)
"
`;

exports[`Table scripts a table definition with no constraints 1`] = `
"CREATE TABLE food (
  id int PRIMARY KEY IDENTITY(1, 1) NOT NULL,
  name nvarchar(255)   NOT NULL,
  inventory_id int   NULL
);
        

"
`;
